### PR TITLE
Improved smart card error handling

### PR DESCRIPTION
**Pull request by josecastillo**
This commit addresses #1303, adding descriptive error messages for the status codes returned by a Yubikey or smart card. A new ```CardException``` class extends ```IOException``` with a field to store the status word returned by the card. Some error conditions should be exposed to the user, like a missing key or a bad PIN; these are implemented with localizable strings that are shown in a notification. Other error conditions are only of interest to developers, like an invalid header byte or command chaining failure. These error conditions should not occur if the application is implemented correctly, and non-localized AssertionErrors are thrown in these cases.

This commit also takes a whack at something I mentioned in #1282: a safety check before the PUT KEY operation. If a user attempts to put a key into a slot that contains a key with a different fingerprint, an IOException is thrown so that the key already on the smart card will not be overwritten.

### CODE REVIEW

**Comment by dschuermann** [comment]
Travis is failing due to my adventure into Android unit tests.
@josecastillo Can you rebase on top of current master? Travis is working again.

**Comment by josecastillo** [comment]
No prob, done and done.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java [hunk 1]
```diff
@@ -105,6 +107,83 @@ public void handleNfcError(IOException e) {
...
+            case 0x6A80:
+                throw new AssertionError("Card returned 'Wrong Data' status; this is a programming error!");
+            case 0x6883:
+                throw new AssertionError("Card expected last command in a chain; this is a programming error!");
+            case 0x6B00:
+                throw new AssertionError("Card reported invalid P1/P2 parameter; this is a programming error!");
+            case 0x6D00:
+                throw new AssertionError("Instruction (INS) not supported by smart card; this is a programming error!");
+            case 0x6E00:
+                throw new AssertionError("Class (CLA) not supported by smart card; this is a programming error!");
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java]: can all of these really never happen due to bad card status or anything? this is a hard crash, some of them sound like they could also happen from a misbehaving smartcard, or simply one that doesn't support openpgp?

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java [hunk 2]
```diff
@@ -105,6 +107,83 @@ public void handleNfcError(IOException e) {
...
+
+            // These errors might be encountered in everyday use, and should display a localized
+            // error message to the user.
+            case 0x6285:
+            {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java]: code style: brace on case line

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java [hunk 3]
```diff
@@ -105,6 +107,83 @@ public void handleNfcError(IOException e) {
...
+                Notify.create(this, getString(R.string.error_nfc,
+                        getString(R.string.error_nfc_terminated)), Style.WARN).show();
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java]: no need for getString here, can pass the resource ids directly

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java (hunk 1 above)
Comment (josecastillo) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java]: This is a good point; I made this assumption from reading through the ykneo-openpgp source code, where they shouldn't happen unless we messed up; but a misbehaving card could force a crash by returning one of these statuses. Plus, someone developing a new NFC-compatible thing could mistakenly crash the app when they're testing it, and an alert would be more useful in that case. I'll turn them into snackbar notifications.

**Comment by Valodim** [comment]
done! thanks for your work! :+1:

**Pull request merged**